homeassistant/components/pi_hole/__init__.py

```diff
@@ -180,26 +108,88 @@ async def async_update_data():
             raise UpdateFailed(f"Failed to communicating with API: {err}")
 
+    entry.add_update_listener(async_update_listener)
+
     coordinator = DataUpdateCoordinator(
         hass,
         _LOGGER,
         name=name,
         update_method=async_update_data,
         update_interval=MIN_TIME_BETWEEN_UPDATES,
     )
-    hass.data[DOMAIN][name] = {
+    hass.data[DOMAIN][entry.entry_id] = {
         DATA_KEY_API: api,
         DATA_KEY_COORDINATOR: coordinator,
     }
 
-    hass.async_create_task(
-        hass.config_entries.async_forward_entry_setup(entry, SENSOR_DOMAIN)
-    )
+    for platform in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
 
     return True
 
 
+async def async_update_listener(hass, entry):
+    """Handle options update."""
+    await hass.config_entries.async_reload(entry.entry_id)
+
+
 async def async_unload_entry(hass, entry):
-    """Unload pi-hole entry."""
-    hass.data[DOMAIN].pop(entry.data[CONF_NAME])
-    return await hass.config_entries.async_forward_entry_unload(entry, SENSOR_DOMAIN)
+    """Unload Pi-hole entry."""
+    hass.data[DOMAIN].pop(entry.entry_id)
+    return all(
+        [
+            await hass.config_entries.async_forward_entry_unload(entry, platform)
+            for platform in PLATFORMS
+        ]
+    )
+
+
+class PiHoleEntity(Entity):
+    """Representation of a Pi-hole entity."""
+
+    def __init__(self, api, coordinator, name, server_unique_id):
+        """Initialize a Pi-hole entity."""
+        self.api = api
+        self.coordinator = coordinator
+        self._name = name
+        self._server_unique_id = server_unique_id
+
+    async def async_added_to_hass(self):
+        """When entity is added to hass."""
+        self.async_on_remove(
+            self.coordinator.async_add_listener(self.async_write_ha_state)
+        )
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return "mdi:pi-hole"
+
+    @property
+    def device_info(self):
+        """Return the device information of the entity."""
+        return {
+            "identifiers": {(DOMAIN, self._server_unique_id)},
+            "name": self._name,
+            "manufacturer": "Pi-hole",
+        }
+
+    @property
+    def available(self):
+        """Could the device be accessed during the last update call."""
+        return self.coordinator.last_update_success
+
+    @property
+    def should_poll(self):
+        """No need to poll. Coordinator notifies entity of updates."""
+        return False
+
+    async def async_update(self):
+        """Get the latest data from the Pi-hole API."""
+        await self.coordinator.async_request_refresh()
+
+    async def async_disable(self, duration=None):
+        """Disable the service for a given duration."""
+        # Do nothing. To avoid problem caused by calling service targeting all entities.
```

Comment:
Why is this a problem? Only the switch platform will register the service so the other entities should not be affected.

Reply:
I don't know if I did anything wrong in registering the entity service. But without this here, I got exceptions saying that
```
AttributeError: 'PiHoleSensor' object has no attribute 'async_disable'
```

Reply:
Hmm, I see now in the entity platform code that that is how it works now, but it's not expected and I don't think intended.

I'll get back to this.

Reply:
We've fixed this now here:
https://github.com/home-assistant/core/pull/37313

If we rebase this PR after that PR is merged we should be able to remove the default method here.

Thanks for bringing this to attention.